server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -112,14 +104,22 @@ void sendErrorResponse(final Version nodeVersion, final TcpChannel channel, fina
     }
 
     private void sendMessage(TcpChannel channel, OutboundMessage networkMessage, ActionListener<Void> listener) throws IOException {
-        MessageSerializer serializer = new MessageSerializer(networkMessage, bigArrays);
-        SendContext sendContext = new SendContext(channel, serializer, listener, serializer);
-        internalSend(channel, sendContext);
+        final BytesStreamOutput bytesStreamOutput = new ReleasableBytesStreamOutput(bigArrays);
+        final ActionListener<Void> wrappedListener = ActionListener.runBefore(listener, bytesStreamOutput::close);
+        final BytesReference message;
+        try {
+            message = networkMessage.serialize(bytesStreamOutput);
+        } catch (Exception e) {
+            logger.error(() -> new ParameterizedMessage("failed to serialize outbound message [{}]", networkMessage), e);
+            wrappedListener.onFailure(e);
+            throw e;
+        }
+        internalSend(channel, message, new SendContext(channel, networkMessage, message.length(), wrappedListener));
```

Comment:
Moar later I think :)

We now always construct the `SendContext` just before calling `internalSend`, may as well construct it inside `internalSend` and set the `startTime` there (so we can make that `final`)

Reply:
Right ... :) actually I went a step further and did away with `SendContext` and the whole `NotifyOnceListener` magic there. No need for that at all, we can simply create a wrapped listener just for the channel send (which should never at this point anyway I think) with it's functionality inline. Let me know what you think.